### PR TITLE
Add match groups and optimize some

### LINKED ISSUE #834: Tilde operator not working as match predicate
**Describe the problem you are trying to solve**

The tilde test expression is documented to work as match predicate:

```
match foo of
  case ~base64|| => "base64"
  default null
end
```

Unfortunately our grammar doesnt support that yet.

**Describe the solution you'd like**
Make the grammar support tilde test expressions.

### PR DESCRIPTION
**Pull request by Licenser**
# Pull request

## Description

This  adds optimization to match statements, they especially target a large number of cases that are quite simple. It also lays the foundation for later optimizations in this space.

## Related

* Related Issues: fixes #834
* [Docs PR](https://github.com/tremor-rs/tremor-www-docs/pull/91)

## Checklist

* [x] The RFC, if required, has been submitted and approved
* [x] Any user-facing impact of the changes is reflected in docs.tremor.rs
* [x] The code is tested
* [x] Use of unsafe code is reasoned about in a comment
* [x] Update CHANGELOG.md appropriately, recording any changes, bug fixes, or other observable changes in behavior
* [x] The performance impact of the change is measured (see below)

## Performance

After discussion, we believe that gaining 200% In complex cases is worth to lose 1 or 2 % of performance on shorter match cases.

### real-workflow-throughput-json
| main | this | delta |
|-|-|-|
|361.6 MB/s|357.3 MB/s| - 1% |

### real-workflow-throughput-json-v0.10
(same logic as `real-workflow-throughput-json` but written in more modern tremor)

| main | this | delta |
|-|-|-|
|368 MB/s|360 MB/s| - 2.2% |

### merge
| main | this | delta |
|-|-|-|
|54.0 MB/s|163.5 MB/s| + 202 % |

### CODE REVIEW

File: packaging/distribution/usr/lib/tremor/tremor.sh [hunk 1]
```diff
@@ -0,0 +1,60 @@
+#!/bin/sh
```
Comment (mfelsche) [inline packaging/distribution/usr/lib/tremor/tremor.sh:1]: why is this file here?

Comment (Licenser) [reply packaging/distribution/usr/lib/tremor/tremor.sh:1]: I noticed it was missing on main :(

File: tremor-script/src/ast/analyzer.rs [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+        let g = if self.guard.is_some() { 10 } else { 0 };
+        g + self.pattern.cost()
+    }
+}
+
+impl<'script> Costly for Pattern<'script> {
+    fn cost(&self) -> u64 {
+        match self {
+            Pattern::DoNotCare | Pattern::Default => 0,
+            Pattern::Expr(_) => 10,
```
Comment (mfelsche) [inline tremor-script/src/ast/analyzer.rs:41]: could you extract those costs to some constants with meaningful names, so the reason for the cost estimation becomes clearer? That would help me and other noobs understand. :)

File: tremor-script/src/ast/analyzer.rs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            ArrayPredicatePattern::Tilde(t) => t.cost(),
+            ArrayPredicatePattern::Record(r) => r.cost(),
+        }
+    }
+}
+
+impl<'script> Costly for ArrayPattern<'script> {
+    fn cost(&self) -> u64 {
+        let s: u64 = self.exprs.iter().map(Costly::cost).sum();
+        s * (self.exprs.len() as u64)
```
Comment (mfelsche) [inline tremor-script/src/ast/analyzer.rs:88]: wouldnt the sum better represent the cost here? After all we iterate over the array pattern and try every match, no? Sounds like a linear thingy.

File: tremor-value/src/value.rs [hunk 4]
```diff
@@ -82,6 +85,170 @@ pub enum Value<'value> {
...
+            (StaticNode::I64(v1), StaticNode::F64(v2)) => {
+                if v2 < i64::min_value() as f64 {
+                    Ordering::Greater
+                } else if v2 > i64::max_value() as f64 {
+                    Ordering::Less
+                } else {
+                    v1.cmp(&(v2 as i64))
+                }
+            }
+            // This is not great!
```
Comment (mfelsche) [inline tremor-value/src/value.rs:150]: why though?

**Review by mfelsche — APPROVED** [review]
This is a magic kolossus! ✨ 

I mostly had questions.

File: tremor-value/src/value.rs (hunk 4 above)
Comment (Licenser) [reply tremor-value/src/value.rs:150]: while we don't expose it but float doesn't implement Ord since it refuses to compare `Nan==Nan` which makes sense so we kind of cheat around it

Comment (mfelsche) [reply tremor-value/src/value.rs:150]: got it.

File: tests/scripts/match_reorder1/script.tremor [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  case %{k ~= glob|eight*|} => 8
+  default => event.k
+end of
+  case a = _ => a
+end;
+
+match a  of
+  case event => event
+  default => "snot"
+end
```
[the remark this answers is not in the document]
Comment (Licenser) [reply tests/scripts/match_reorder1/script.tremor:62]: :+1:

File: tremor-script/src/ast.rs [hunk 6]
```diff
@@ -1363,10 +1751,40 @@ pub enum Pattern<'script> {
...
     }
+    fn is_exclusive_to(&self, other: &Self) -> bool {
+        match (self, other) {
+            // Two literals that are different are distinct
+            (Pattern::Expr(ImutExprInt::Literal(l1)), Pattern::Expr(ImutExprInt::Literal(l2))) => {
+                l1 != l2
+            }
+            // For record patterns we compare directly
+            (Pattern::Record(r1), Pattern::Record(r2)) => {
+                r1.is_exclusive_to(r2) || r2.is_exclusive_to(r1)
```
[the remark this answers is not in the document]
Comment (Licenser) [reply tremor-script/src/ast.rs:1740]: So we can implement exclusivity in one direction instead of both, for example for `extractors` and `==` we only implement the `does an extractor is exclusive to a ==` not `is a == exclusive to an extractor`.

That said I'm not 100% sure where the 'right' place is to do this doubling :( it is in more places then I liked

File: tremor-script/src/ast.rs [hunk 7]
```diff
@@ -1435,6 +1821,91 @@ pub enum PredicatePattern<'script> {
...
+                    lhs: lhs1,
+                    kind: BinOpKind::Eq,
+                    ..
+                },
+                PredicatePattern::Bin {
+                    lhs: lhs2,
+                    kind: BinOpKind::Eq,
+                    ..
+                },
+            ) if lhs1 == lhs2 => true,
```
[the remark this answers is not in the document]
Comment (Licenser) [reply tremor-script/src/ast.rs:1837]: good catch yes! even so having the same RHS might be silly it is something we should not change the behaviour on!

File: tremor-script/src/ast.rs [hunk 8]
```diff
@@ -1140,59 +1212,373 @@ pub struct TestExpr {
...
+        }
+    }
+
+    fn replace_last_shadow_use(&mut self, replace_idx: usize) {
+        match self {
+            Self::Simple { patterns, .. } => {
+                for p in patterns {
+                    let mut e = Ex::null_lit();
+                    mem::swap(&mut p.last_expr, &mut e);
+                    p.last_expr = e.replace_last_shadow_use(replace_idx);
```
[the remark this answers is not in the document]
Comment (Licenser) [reply tremor-script/src/ast.rs:1372]: That's what I wanted to write, sadly rust says no :(

File: tremor-script/src/ast/analyzer.rs (hunk 3 above)
Comment (Licenser) [reply tremor-script/src/ast/analyzer.rs:88]: for each pattern in the array we itterate over the entire target array, so 3 patterns on a 10 element array iterate over 30 elements. We don't know how many elements will be there in the final array but a conservative approximation is that if you write `n` different patterns the array you're looking at will have at least `n` elements.

PS: This should be in a comment :D (will copy it in.

File: tremor-value/src/value.rs (hunk 4 above)
Comment (Licenser) [reply tremor-value/src/value.rs:150]: added this as a comment

File: tremor-script/src/ast/analyzer.rs (hunk 2 above)
Comment (Licenser) [reply tremor-script/src/ast/analyzer.rs:41]: not sure how much it'll help but I'll add some, at least it'll make adjusting them easier later :D :+1:

File: tremor-script/src/ast.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (mfelsche) [reply tremor-script/src/ast.rs:1752]: i think it is ok. We can revisit this when needed. I cannot come up with a case where the exclusivity relation is not commutative. So if anyhow, this might only be a problem of doing something twice, but thats ok as we are not in the hot path.

File: tremor-script/src/ast/analyzer.rs (hunk 3 above)
Comment (mfelsche) [reply tremor-script/src/ast/analyzer.rs:111]: Got it. I mixed it up with the tuple pattern logic. All good

**Review by mfelsche — APPROVED** [review]
Love it!

**Comment by Licenser** [comment]
Question: are we OK with a 2-3% perf drop on the baseline test for the perf gain on more complex match statements?

**Comment by mfelsche** [comment]
the baseline test is a single match arm? or no match at all? Is the performance drop really attributable to your changes? How is the behaviour with middle-sized match statements, say around 10 arms?

**Comment by Licenser** [comment]
https://github.com/tremor-rs/tremor-runtime/blob/main/bench/real-workflow-throughput-json/main.trickle is the script in question, it is written in a fairly outdated style (we had no pattern matches back then) I'll make a `new` variant to see how that would fare

**Comment by Licenser** [comment]
I'm not particularly happy with the decrease in performance for `real-workflow-throughput-json`

**Review by mfelsche — APPROVED** [review]
Finally!
